### ISSUE TITLE
Output file naming from ODDs

### ISSUE THREAD

**Issue by martindholmes** [TEIC/Stylesheets#201]
Iif you run the teitodtd command-line script, you will get:

myfile.odd.dtd

The extra ".odd" is unnecessary, and @sydb and I think it should be simply myfile.dtd.

Also, if you use teitorelaxng, you get:

myfile.odd.relaxng

This should be .rng.

**Comment by hcayless** [comment]
Not objecting to this, but you *can* specify the output file name. 

Having the default output file name be a bit weird might help avoid accidentally stomping on an existing myfile.rng, so that might be why it is the way it is. It also encourages you to be explicit.

**Comment by lb42** [comment]
I agree with Hugh.

**Comment by martindholmes** [comment]
The Oxygen tasks create a folder called "out" and put the results there; that's presumably also to avoid overwriting something by accident. If we agree that some such strategy is a good idea, I think they should both employ the same strategy.

**Comment by sydb** [comment]
I have to admit that when @martindholmes and I were talking about this earlier I had forgotten that you can specify an output file name. That may change my calculus on this. But certainly @martindholmes has a point, it would be nicer if the defaults were the same for the commandline `Stylesheets/bin/teitoXXX` and the transformation scenarios in our oXygen framework.

**Comment by lb42** [comment]
Removing the bug label. It's not a bug. It's a feature that needs discussion!

**Comment by martindholmes** [comment]
Council believes that the command-line process should be brought into line with what happens in Oxygen, for consistency, so files should be generated in an "out" folder.

**Comment by jamescummings** [comment]
What happens when an out folder already exists with identical filenames? (i.e. I've done this transformation, found something I forget to change, changed it, then done the transformation again.)  
Does it overwrite?
Does it call the files something else?
Does it fail?

**Comment by scstanley7** [comment]
Council seems to be changing their minds on this: We don't agree about the "out" bit, so we won't tackle that right now.

We *do* agree that the file extensions should be consistent across (i.e. change ".relaxng" to ".rng" consistently), and we agree that we should have "myfile.odd" transform to "myfile.rng" instead of "myfile.odd.rng"

**Comment by martindholmes** [comment]
@sydb and I created a new branch https://github.com/TEIC/Stylesheets/tree/issue-201 for this, but we're still in the planning stages. We note:

 1. Council would prefer the Stylesheets defaults to match those of the Oxygen plugin, so output should end in out/filename.ext rather than in filename.inputExt.outputExt.

 1. The symlink system, whereby the input file type and the output file type are derived from the name of the symlink itself, is very elegant, but if you don't explicitly provide an output filename, you end up with silly extensions like .markdown and .relaxng; we would prefer to get the default output extension from a lookup table that needs to be constructed in the transformtei script, so that although you call e.g.
`teitomarkdown mytei.xml`

you get:
`mytei.md`

rather than:
`mytei.xml.markdown`

In other words, the script would derive the output format from the symlink name, but then look it up to find that the proper output extension is `.md`, and then construct the output filename _without_ retaining the input file extension.

 1. We also believe that there should be a deprecation mechanism whereby a warning is issued on the command line for any symlink whose behaviour has changed, explaining the change, so that people who have used the symlinks in their scripts will know why things break.

1. OxGarage will also have to be checked to see whether it's using any of this stuff. We don't know whether it is or not; it may just be using the Stylesheets directly. If its behaviour will change, some explanatory warnings should be added.

1. We disagree about whether the default extension for a TEI output file should be `.tei` (as currently) or `.xml`. Council should presumably weigh in on this. One view holds that we should encourage the use of the `.tei` extension alongside other extensions such as `.xsl` `.xspec`; the other maintains that the vast majority of TEI files in the world have the `.xml` extension, so we'll be annoying more people if we output `.tei`.

1. One mechanism we have discussed for a warning system is to leave the old symlink become a script which issues the warning, then calls the new symlink. So `teitorelaxng` would warn that the default output filename has changed from `file.inputExt.relaxng` to `file.rng`, and then call teitorng (a new symlink).

**Comment by sydb** [comment]
@martindholmes (and everyone else):
On Friday we missed [@scstanley7’s comment](#issuecomment-419608527) above that points out Council is not firmly in the out/ camp. Thus (1) above may need to be adjusted.

**Comment by martindholmes** [comment]
@sydb I think we need to aim for a change which will cause the lowest probability of unexpected and potentially damaging behaviour for people who may not be expecting the change, so I would support using the `out` folder that is used in the Oxygen plugin for the command line as well, much as I dislike it myself.

**Comment by ebeshero** [comment]
I agree with @martindholmes here that we should preserve the usual output behavior and file in `out/`.

**Comment by sydb** [comment]
Not that I think there is a lot of stake in this, but one counter-argument is that changing the behavior of a necessarily somewhat manual process (oXygen) is a lot less impactfull than changing the beavior of a process that may be close to fully automated (the scripts).

Another is that I would not be surprised if most users either don’t care or would prefer to ditch the out/ directory, even if it means a change.

**Comment by lb42** [comment]
If you don't like using "out" you can of course also change that default behaviour in your local oXygen installation

**Comment by martindholmes** [comment]
@lb42 Yes, one of the first things I do in a new Oxygen setup is to duplicated that scenario and change it to remove the out folder. But that requires a little understanding of how Oxygen transformation scenarios work, which not everyone has. @sydb I suspect that if people are running these things through scripting, they're already providing their own output file as a param, so they're less likely to be affected, but of course all this is guesswork. It's likely we'll upset some small number of people whatever we do, but I think it's better in the long run if we provide consistent behaviour however you invoke the Stylesheets transformations.

**Comment by sydb** [comment]
@martindholmes, as I suspect you know, I agree strongly with the idea that there should be consistent behavior from the Stylesheets transforms, and am moderately in favor of dropping the out/ directory.

**Comment by martindholmes** [comment]
@sydb @lb42 @martinascholger @hcayless @ebeshero I don't believe there will ever be consensus on this, so I think this has become a zombie ticket. I think we should close it.

**Comment by martindholmes** [comment]
@HelenaSabel, @sydb, @npcole and I believe this ticket should be split into two new tickets: one is the question of the /out folder, which needs Council consensus, and the other is the question of the multiple extensions (file.odd.dtd, file.odd.relaxng), which is pernicious because, for one thing, on Windows, which hides the actual extension by default, it will appear that file.odd.dtd is actually file.odd. We will close this one and raise the two new ones.

**Comment by martindholmes** [comment]
Issue #543 is the successor issue which addresses multiple extensions.
